### PR TITLE
Revert to 1.3 + cherry-picks

### PR DESCRIPTION
**Pull request by jmesnil**
As discussed in the meeting today, we agreed to revert to 1.3 in order to have a stable releasable master branch and add new feature one at a time with appropriate use cases / tests / spec.

I did not want to lose any changes we did after 1.3 that consolidates the API and clarified the spec so what I did was
1. Revert to 884c1d30f3c7a13440e981ad00556833574fba8c (the maven commit to prepare 1.4-SNAPSHOT) just after 1.3 tag[1]
2. cherry-pick any commit after 1.3 that clarified the spec / API without introducing any functional changes.
3. I added a commit at the end to bring back some stuff from the reverted commits that were still interesting to keep (javadoc changes, better sentences in the spec)

In addition to the diff from master, another diff to look at is the [diff to 1.3](https://github.com/eclipse/microprofile-config/compare/1.3...jmesnil:revert_to_1.3) which shows what will actually be going in the new release of MicroProfile Config

[1] I used these commands to do this revert commit:

```
git reset --hard 884c1d30f3c7a13440e981ad00556833574fba8c
git reset --soft HEAD@{1}
git commit
```

### CODE REVIEW

**Comment by jmesnil** [comment]
@Emily-Jiang the ECA tool is complaining about missing sign-offs but I only cherry-picked from commits that were already in the project

**Comment by Emily-Jiang** [comment]
@jmesnil You need to fix your ECA. Have you signed the new version of agreement?

**Comment by Emily-Jiang** [comment]
I had a 2nd thought. How best to preserve the ConfigAccessor changes with the original contribution as it is very likely we want to add back in? Will it be easily just deleting ConfigSnapshot and cacheFor? Thoughts?

**Comment by dmlloyd** [comment]
> I had a 2nd thought. How best to preserve the ConfigAccessor changes with the original contribution as it is very likely we want to add back in? Will it be easily just deleting ConfigSnapshot and cacheFor? Thoughts?

Definitely in a separate PR.    See #445 for one such option.

**Comment by jmesnil** [comment]
@Emily-Jiang The ECA tool has issues with other sign-offs than mine (see https://accounts.eclipse.org/legal/eca/validation/13334).

I think some commits I cherry-picked were directly applied to master and then were not validated by the tool. But all these commits are already in the project history so I see no issue with merging them again

**Comment by dmlloyd** [comment]
This fixes #442.

**Comment by dmlloyd** [comment]
IMO this should be merged ASAP.  @Emily-Jiang WDYT?

File: api/src/main/java/org/eclipse/microprofile/config/Config.java [hunk 1]
```diff
@@ -121,77 +113,20 @@
...
-     * @param configValues the list of {@link ConfigAccessor} to be accessed in an atomic way
-     *
-     * @return a new {@link ConfigSnapshot} which holds the resolved values of all the {@code configValues}.
-     */
-    ConfigSnapshot snapshotFor(ConfigAccessor<?>... configValues);
-
-
-    /**
-     * Return all property names used in any of the underlying {@link ConfigSource ConfigSources}.
+     * Return a collection of property names.
```
Comment (Emily-Jiang) [inline api/src/main/java/org/eclipse/microprofile/config/Config.java]: I prefer the previous line.

File: api/src/main/java/org/eclipse/microprofile/config/inject/ConfigProperty.java [hunk 2]
```diff
@@ -137,25 +136,4 @@
      */
     @Nonbinding
     String defaultValue() default UNCONFIGURED_VALUE;
-
```
Comment (Emily-Jiang) [inline api/src/main/java/org/eclipse/microprofile/config/inject/ConfigProperty.java]: In this file, the contributor Tomas should be removed as well.

**Comment by Emily-Jiang** [comment]
In the error report, it says: Anonymous (jmesnil@****.com) is not covered by the necessary legal agreements in order to proceed. Anyway, can you look at your branch and remove the contributors that committed to the deleted code?

**Comment by jmesnil** [comment]
@Emily-Jiang I fixed the contributors list.

Re the ECA tool, the failure for "Anonymous (jmesnil@****.com)" is caused by a direct commit I did by mistake on GitHub UI (I thought that it'd create a PR but merged to master directly instead): https://github.com/eclipse/microprofile-config/commit/3c04b3a430543643d0ef2df0b0dde8cbad17b980

Likewise, the error for you is related to https://github.com/eclipse/microprofile-config/commit/624426937d1b1f7a5704115024cb411e9cc05b73 that was not signed-off

(we should rely configure GitHub to prevent any direct commit to master to avoid these errors)

**Comment by jmesnil** [comment]
> I had a 2nd thought. How best to preserve the ConfigAccessor changes with the original contribution as it is very likely we want to add back in? Will it be easily just deleting ConfigSnapshot and cacheFor? Thoughts?

There is also variableExpansion and attribute change support, etc. that are still not specified/test properly.

Going back to 1.3 + clarifications allow us to move forward again by adding feature one at a time with a clear understanding and agreement on them.

Keeping master and deleting bits here and there would add entropy to the current state that is not stable enough.
My intent with that PR is to be stable (and releasable again) so that new features can be then added properly

**Comment by dmlloyd** [comment]
As I've said in other contexts, the `ConfigAccessor` I present as an option is almost completely different to the one that was originally in the `master` branch.

I am however OK with the idea of bringing over the original copyright and authors tags.

**Comment by Emily-Jiang** [comment]
As today's hangout, we agreed to merge this PR. The ECA validation errors displayed were the removed contributions.

**Pull request merged**